Add `allowedButtonsFilter` to prevent Draggable from appearing with secondary click.

bernaferrari opened the pull request:
This changes the default Draggable behavior without breaking any test. It _doesn't need_ to change the default behavior, but I think "left click" should be the default to prevent interfering with right-click and scroll-click. This is specially important given people using right click to show options. So, the current behavior feels like a bug.

`allowListPointer` is a new parameter:
- null (allow every value, current behavior)
- [] (allow no value)
- [kPrimaryButton] (what most people will use).

Fixes: https://github.com/flutter/flutter/issues/111850
Fixes: https://github.com/flutter/flutter/issues/89962

a-wallen commented:
👋  @bernaferrari thanks for the contribution. I'm going to try my best to help get this one across the finish line. I think that we definitely want this PR, but we also want to do it the _right_ way. I'm going to break up my feedback into 3 parts.

1. Revisions regarding changes to the API design.
2. Revisions regarding the functionality of the code.
3. Revisions regarding the readability of the code.

a-wallen commented on packages/flutter/lib/src/widgets/drag_target.dart:
```diff
@@ -179,6 +179,7 @@ class Draggable<T extends Object> extends StatefulWidget {
     this.ignoringFeedbackPointer = true,
     this.rootOverlay = false,
     this.hitTestBehavior = HitTestBehavior.deferToChild,
+    this.allowListPointer = const <int>[kPrimaryButton]
```
Design: If `this.allowListPointer` now has a default value, then all clients of this API will be opted-in. I think that we eventually want to do this, but not yet. 

[Flutter's compatibility policy](https://docs.flutter.dev/resources/compatibility) does not qualify this as a breaking change because it does not break existing tests. However, I think that we should treat it as such since the behavior of apps will change.

For now, I suggest that we allow this value to be null and make a new issue to track widgets where we make similar changes. How does that sound?

- [ ] Action Item: Make the following change.
```dart
this.allowListPointer,
```

a-wallen replied:
Another idea here is that the framework should allow the user to determine the predicate (what is allowed) for the `PointerDownEvent`. 

```dart
// default the value in the parameter list
// we default it here so that our changes are backward compatible
this.isPointerAllowed = (PointerDownEvent event) => true,

// where it is defined as a property in the class
final bool Function(PointerDownEvent)? isPointerAllowed;
```

Right now the way that this change checks whether we allow this PointerDownEvent is 
```dart
if (widget.allowListPointer == null || widget.allowListPointer!.contains(event.buttons))
```
but what if the `PointerDownEvent` contains both `kPrimaryButton & kSecondaryButton`? 

The above code suggests that this gesture is allowed, which is heavy-handed considering that users might want to ignore that. 

This example should illustrate how using a predicate might be more advantageous rather than an allow list.

bernaferrari replied:
> For now, I suggest that we allow this value to be null and make a new issue to track widgets where we make similar changes. How does that sound?

Yes, I can make this. But... I don't think any other widget is affected. Sure, it might be, I agree we can study about this, but from my head, most of them use GestureDetector or something else. This one is using bare-bones Listener, so it behaves as an exception.

bernaferrari replied:
> what if the PointerDownEvent contains both kPrimaryButton & kSecondaryButton?

There is a test for this. The allowList allows anything inside. If you have kPrimary and kSecondary, it would work for any of them. But would fail for Scroll-click. If you had [kPrimary, kSecondary, kScrollClick(?)], then it would behave like right now.

bernaferrari replied:
This is more radical, in Listener:
![image](https://user-images.githubusercontent.com/351125/191351885-ea50b906-5a57-44d4-8d03-3541fa44510e.png)

bernaferrari replied:
These are all the Listener usages inside /widgets. Most are related to scrolling.

<img width="618" alt="image" src="https://user-images.githubusercontent.com/351125/191354020-b4c6b227-5db9-4bee-953f-222d19c0460e.png">
<img width="694" alt="image" src="https://user-images.githubusercontent.com/351125/191354097-835acd57-7033-4a09-96e5-ffd466c2179d.png">

<img width="484" alt="image" src="https://user-images.githubusercontent.com/351125/191354558-71a469d3-e987-4292-8315-77a74638b387.png">

a-wallen replied:
> Another question I have is: what if we "fixed" Listener to only allow specific buttons?

That's a more impactful solution that we should only take if it's absolutely necessary. If you can justify a reason why it should be addressed there, we can consider it. Otherwise, we should narrow the scope of handling `PointerDownEvents` with the "right-click only" behavior to `Draggable.`

a-wallen replied:
> Yeah, I'm not sure I like the predicate thing, probably the allow list concept is easier to understand.

I'll give more context and an elaborate example. Let me know if this helps.

## From [`PointerDownEvent.buttons`](https://api.flutter.dev/flutter/gestures/PointerEvent/buttons.html)  
> Bit field using the *Button constants such as [kPrimaryMouseButton](https://api.flutter.dev/flutter/gestures/kPrimaryMouseButton-constant.html), [kSecondaryStylusButton](https://api.flutter.dev/flutter/gestures/kSecondaryStylusButton-constant.html), etc.
> 
> For example, if this has the value 6 and the [kind](https://api.flutter.dev/flutter/gestures/PointerEvent/kind.html) is [PointerDeviceKind.invertedStylus](https://api.flutter.dev/flutter/dart-ui/PointerDeviceKind.html), then this indicates an upside-down stylus with both its primary and secondary buttons pressed.

`PointerDownEvent.buttons` can be composed of many buttons. 

Now, suppose that we just have two buttons. For example, `kPrimaryButton = (0x01)/(0001)` and `kSecondaryButton = (0x02)/(0010)`. If the `PointerDownEvent` registered a click from `kPrimaryButton` and `kSecondaryButton` its value will be `kPrimaryButton + kSecondaryButton = 4`. If our allowlist values contains `[kPrimaryButton, kSecondaryButton]`

This condition 
```dart
widget.allowListPointer!.contains(event.buttons))
```
is false since `[1, 2]` does not contain `3`. But, I think you mean for it to be true.

bernaferrari replied:
Oh, crap. I hate these. I thought they were enum (maybe we can make them enhanced enum now? 👀). You are right, the allowList wouldn't work for this.
But `kPrimary in [kPrimary, kSecondary]` would work, just not `kPrimary` in `kPrimary + kSecondary`.

bernaferrari replied:
So, the allowList could be `int? allowButtons` How do you feel? 0 = none. Null = all.

a-wallen replied:
> So, the allowList could be int? allowButtons How do you feel? 0 = none. Null = all.

That's definitely another approach that we could take here, but that would defer the calculation of `allowButtons` via bitwise operations to the client of this API. The goal here should be to make the API as simple as possible to use, and I think we have an opportunity here to hide some of the messy details of `PointerDownEvent.buttons` and programmability via something like `allowList` accessible to beginners.

bernaferrari replied:
Thanks for all your helpful comments. I just pushed some great changes. Now I'll see how far I can get with the bitwise you shared with me 😀

bernaferrari replied:
Your algorithm works!!!! That's fantastic. lol.

bernaferrari replied:
I need to work now. I'll check what's wrong with the tests tomorrow. Feel free to check if you like the current state :)

See you later! Thanks for all the help!

a-wallen commented on packages/flutter/lib/src/widgets/drag_target.dart:
```diff
@@ -359,6 +360,14 @@ class Draggable<T extends Object> extends StatefulWidget {
+  /// When value is null, allow all interactions.
+  ///
+  /// Defaults to [kPrimaryButton].
+  final List<int>? allowListPointer;
```
If I'm not mistaken, when the previous value defaulted,
```dart
this.allowListPointer = const <int>[kPrimaryButton]
```
this value cannot be null.

bernaferrari commented:
Sorry, I thought this would be my simplest PR in a while. I didn't expect that much complexity. After talking with the other reviewer (I won't tag, so he isn't notified), he gave the idea to modify MultipleGestureDetector and I like the idea. Just committed did.

HansMuller commented:
It would be good to apply this change to all of the components where it's warranted, not just Draggable. Like scrollbar thumbs, slider thumbs, tab dragging, there are probably others.

moffatman commented:
I thought it wouldn't apply to all `GestureRecognizer` subclasses, since some have specific handling and separated callbacks for different buttons already. And it looks like the single-drag recognizers that are more commonly used (`HorizontalDragGestureRecognizer` etc) already only allow drags with the primary button.

Piinks commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -223,6 +223,7 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
     )
     super.kind,
     super.supportedDevices,
+    this.allowedMouseButtons,
```
This seems to conflict with the `supportedDevices` parameter above it. What if it does not include `PointerDeviceKind.mouse`?

bernaferrari replied:
That's a good question. Shall I make `supportedDevices == mouse && ...`?

Piinks replied:
I am not sure. In general, having both parameters may be a bit unclear as far as the API surface goes.

What about other gestures? Or other places in the framework where we handle mouse buttons? It feels like we might need a more abstract design that can satisfy a consistent pattern across the framework.

bernaferrari replied:
Yeah, I thought about this. The `buttons` API is also "very weird" because it uses bit-wise operations while everything else uses enum + set.

I just, well... don't know how to propose something better. Ideally it would be a sub-property from mouse, so you could have `mouse(clickType: primary | secondary | scroll | others)`, then unfold the mouse property when mouse is being used, but I don't think Dart supports anything like this (seems more like a sealed class thing?).

moffatman replied:
You can right-click with `stylus`, and in some cases `touch` depending on iPadOS Info.plist. Probably the field name should be changed to `allowedButtons`. But on `Draggable` it could be kept as `allowedMouseButtons`, since in that context `allowedButtons` makes me think of a button widget, and since the parameter type is `int`, it is not really self-documenting.

bernaferrari replied:
`typedef mouseButtons = int`. A possible way that could make allowedMouseButtons easier to understand.

Piinks replied:
Making this specific to mice excludes the other valid inout devices listed here, like stylus. What about `inputDeviceModifiers` or something along those lines? Keeping it abstract will be inclusive of all input devices, and avoid confusion with the term `buttons`.

bernaferrari replied:
Sure, we could have something like that. But `allowedMouseButtons` that's executed from behind the scenes is very "obscure"/low-level. It uses the bit-wise operations and so... But I'll try a better name, let's see.

bernaferrari replied:
I changed from `allowedMouseButtons` to `allowedInputPointers`, but Flutter uses buttons in many places, like here:
<img width="573" alt="image" src="https://user-images.githubusercontent.com/351125/194918169-8f0eadb1-4137-429c-9084-ca2087f5de5f.png">

bernaferrari commented:
Here is what Flutter does.

`DragGestureRecognizer` (slider) is currently locked to primary button. `MultiDragGestureRecognizer` doesn't lock to anything (this proposal).

`MultiTapGestureRecognizer`, `ForcePressGestureRecognizer` and `ScaleGestureRecognizer` *could* have unexpected behaviors since they don't check for primary button. Someone said it is possible to have right click on iPad with touch.

<img width="941" alt="image" src="https://user-images.githubusercontent.com/351125/192175042-66c407b5-4c39-4d4c-939e-3eac3fa854ca.png">

bernaferrari commented:
I renamed it again, and fixed the tests. If/after this lands, we could update DragGestureRecognizer (sliders) to do the same, so they would share a similar API.

a-wallen commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -328,6 +339,18 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
+  @protected
+  bool isPointerAllowed(PointerDownEvent event) {
+    // Check for mouse button, then device kind.
+    if (inputPointerBehavior != null) {
```
First of all, good job addressing the feedback from this PR, it has evolved a lot since the first draft.

For the line above, I'm wondering why we aren't assigning `(event.buttons & kPrimaryButton != 0)` as the default value for `inputPointerBehavior`.

If `inputPointerBehavior` defaults to `(event.buttons & kPrimaryButton != 0)`, the code below can be simplified:
```dart
  /// Checks whether or not a pointer is allowed to be tracked by this recognizer.
  @override
  @protected
  bool isPointerAllowed(PointerDownEvent event) {
    return inputPointerBehavior!(event.buttons) && super.isPointerAllowed(event); 
  }
```

moffatman replied:
Why was this closed? I had the same thought.

bernaferrari replied:
You can't have a const function constructor, so the only option is: `inputBehavior?.call(event.buttons) ?? (event.buttons & kPrimaryButton != 0)`

moffatman replied:
You could have the default behavior as a private function in the file. Like above all the recognizers:

```dart
bool _defaultButtonAcceptBehavior(int buttons) => buttons & kPrimaryButton != 0;
```

And this could be used in the constructor as the default value.

bernaferrari replied:
Can it?
<img width="1073" alt="image" src="https://user-images.githubusercontent.com/351125/196819979-a7921e80-ed69-4adf-bac9-6e2bda2c8ec0.png">

bernaferrari replied:
OHHH, GOT IT!
TIL something new!

bernaferrari replied:
That's awesome! Thanks!

bernaferrari replied:
For monodrag (future PR), do you think I should make this function public, or copy and paste the same thing in `monodrag.dart`?

moffatman replied:
I think it's better to copy and paste the same thing, to avoid polluting the global namespace for anyone who imports `package:flutter/gestures.dart`.

a-wallen reviewed (approved):
This change LGTM, but there should be an issue to update DragGestureRecognizer before we send this off.

bernaferrari commented:
cc @moffatman after many many many comments and changes, now is your time to review! 🎉

moffatman commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -231,6 +232,16 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
+  /// When value is `(int buttons) => buttons & kPrimaryButton != 0`, allow
+  /// if selection contains a primary button.
+  ///
+  /// Defaults to `(int buttons) => buttons & kPrimaryButton != 0`.
```
I think the default value should be explained rather than showing the code, especially since it involves bitwise operations.

bernaferrari replied:
Thanks. I made the default `buttons == kPrimaryButton` to simplify things (the other reviewer agrees). I'll write what's possible to do with bitwise.

moffatman commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -231,6 +232,16 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
+  /// if selection contains a primary button.
+  ///
+  /// Defaults to `(int buttons) => buttons & kPrimaryButton != 0`.
+  final bool Function(int buttons)? inputPointerBehavior;
```
Can you add a typedef for this so we can have a name for the predicate?

bernaferrari replied:
Yes, or pass the pointer and call `pointer.buttons` so the user has more data.

a-wallen replied:
[Style Guide](https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#naming-rules-for-typedefs-and-function-variables): consider adding a typedef. This might improve the readability.

a-wallen replied:
`isPointerAllowed` sounds good to me. https://stackoverflow.com/questions/8548232/whats-the-best-way-to-name-a-predicate-function

moffatman commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -223,6 +223,7 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
     )
     super.kind,
     super.supportedDevices,
+    this.inputPointerBehavior,
```
inputPointerBehavior so generic, it really could refer to so many things. Maybe something like `isButtonAllowed`, `areButtonsAllowed`, or `buttonAcceptBehavior`?

bernaferrari replied:
I agree. Source: https://github.com/flutter/flutter/pull/111852#discussion_r986298568.

I'll try one of the names you suggested.

bernaferrari replied:
The hard challenge is "button" doesn't convey a lot of meaning.

moffatman replied:
This is in the `GestureRecognizer` level, not the framework / `GestureDetector` level. So I think button is pretty clearly relating to a mouse (pointer) button rather than some widget. Maybe specifically using button**s** instead of button will associate it more with the `buttons` field on `PointerEvent`?

a-wallen replied:
`isPointerAllowed` sounds good to me. https://stackoverflow.com/questions/8548232/whats-the-best-way-to-name-a-predicate-function

Piinks reviewed (commented):
I am still working through reviewing this myself, but I just wanted to quickly contribute based on the conversation happening that I really don't think this should include `buttons`. In the context of Draggable, and all of the other classes this would extend to (think all of the classes that have `supportedDevices`, there are ~27) I don't know that it is very clear what it is supposed to be if it is a `button`.

I should be able to finish this review tomorrow.

bernaferrari commented:
I agree `buttons` is not clear, but everybody is struggling to find a better name (specially since the internal methods use buttons).

The callback is `int buttons` right now. Could be `PointerEvent pointer`. We are one step from just transforming `isPointerAllowed` into a builder.

<img width="535" alt="image" src="https://user-images.githubusercontent.com/351125/196826182-d015dc9c-0538-4329-ad0c-82287c748ee7.png">

Piinks commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -231,6 +236,22 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
+  /// When value is `(int buttons) => true`, allow all interactions.
+  ///
+  /// Defaults to [_defaultButtonAcceptBehavior].
+  final bool Function(int buttons) inputPointerBehavior;
```
This should have a typedef

bernaferrari replied:
Thanks, done :)

Piinks commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -231,6 +236,22 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
   /// [Drag] object returned by this callback.
   GestureMultiDragStartCallback? onStart;
 
+  /// Called when interaction starts. This limits the dragging behavior
```
Can you put this doc in a template and reuse below?

bernaferrari replied:
Yes, but how? I never did this. lol
Do you have any example/reference I can take a look?

bernaferrari replied:
Thanks, done :)

Piinks commented on packages/flutter/lib/src/widgets/drag_target.dart:
```diff
@@ -359,6 +360,16 @@ class Draggable<T extends Object> extends StatefulWidget {
   /// Defaults to [HitTestBehavior.deferToChild].
   final HitTestBehavior hitTestBehavior;
 
+  /// Called when interaction starts. This limits the dragging behavior
```
The templated doc can be re-used here.

bernaferrari replied:
Done :)

Piinks reviewed (commented):
Ah! I see now. `button` makes a lot more sense with `PointerEvent.buttons`. I think that is what I was missing. Can you add reference in the docs here to `PointerEvent.buttons`?

a-wallen commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -26,6 +26,10 @@ export 'gesture_settings.dart' show DeviceGestureSettings;
 
+/// The default value for inputPointerBehavior.
+/// Accept the input if, and only if, [kPrimaryButton] is pressed.
+bool _defaultButtonAcceptBehavior(int buttons) => buttons == kPrimaryButton;
```
I think that this can be a static variable.

bernaferrari replied:
Can't:
https://dart-lang.github.io/linter/lints/prefer_function_declarations_over_variables.html

a-wallen replied:
Sorry, that's my bad... it wasn't a variable, to begin with. Regardless, this function declaration can still be static if I am not mistaken.

moffatman replied:
Ah I agree, I think it could probably be a static method within the drag recognizer class

bernaferrari replied:
Done :)

bernaferrari commented:
Testing a new name: `InputButtonsFilter inputButtonsFilter`. The review comments are fixed, documentation is improved.

goderbauer commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -26,6 +26,11 @@ export 'gesture_settings.dart' show DeviceGestureSettings;
 /// Signature for when [MultiDragGestureRecognizer] recognizes the start of a drag gesture.
 typedef GestureMultiDragStartCallback = Drag? Function(Offset position);
 
+/// Signature for [MultiDragGestureRecognizer._buttonsEventFilter].
```
fly-by: Please don't reference private types in API docs.

bernaferrari replied:
Thanks! Done.

Piinks commented:
There has been a lot of really great collaboration on this PR, although I feel that it is not super clear the best way to proceed. Everyone's input is super valid, and I think it is important to still consider the framework at large and having a consistent pattern that works across multiple use cases before we commit to one.

@Renzo-Olivares has been working in this area as well, maybe he has some suggestions on how we can approach this thinking about framework mouse gestures as a whole, and not just one subset.

Renzo-Olivares reviewed (commented):
Sorry it took me a bit to catch up to the conversation. 

This is an awesome contribution! It adds a very nice convenience for something that was previously verbose to do. It feels like it goes hand in hand with `supportedDevices`.

I think currently the only alternative to accomplishing what this PR does is extending the desired recognizer like below

```dart
class OverridingAllowedPointerHorizontalMultiDragGestureRecognizer extends HorizontalMultiDragGestureRecognizer {
  @override
  bool isPointerAllowed(PointerEvent event) {
    return true;
  }
}
```

In the case of `Draggable` that uses 3 types of `MultiDragGestureRecognizers`, then we would have to extend all 3 which is a bit verbose. Having a parameter for this is much cleaner.

I agree with the sentiment that we want to make this pattern consistent across the entire framework. So eventually we should add this parameter to `GestureRecognizer` to give all recognizers a way to choose there allowed pointers. I think it's okay to open up an issue for that and do it in a separate PR. It's also important to note that there are some `asserts` around some recognizers that check for some private `_initialButtons` variable, such as in `DragGestureRecognizer` which checks for `kPrimaryButton`. We will most likely want to change or remove those.

bernaferrari commented on packages/flutter/lib/src/gestures/multidrag.dart:
```diff
@@ -223,14 +228,38 @@ abstract class MultiDragGestureRecognizer extends GestureRecognizer {
+  /// {@endtemplate}
+  final ButtonsEventFilter _buttonsEventFilter;
+
+  /// The default value for [_buttonsEventFilter]..
```
Like this? It breaks the [] reference, but it is kind of what we want.
<img width="797" alt="image" src="https://user-images.githubusercontent.com/351125/200417602-302ad1dd-2d0d-4009-b816-4030124f7e05.png">

bernaferrari replied:
Ok, I did it. If you could re-review, I would appreciate it.

Renzo-Olivares replied:
Sorry about that, yes I can give this another look.

Renzo-Olivares reviewed (changes requested):
LGTM other than my two comments. Sorry for the delay on reviewing this one. I think this PR is an awesome addition to the framework!

I do want to make sure that the API stays consistent between gesture recognizers. @bernaferrari could you open an issue with something like below with some added context. You don't have to assign it to yourself unless you want to work on it.

What do you think about this plan?

GestureRecognizer `buttonsEventFilter` API support project :umbrella:
- [ ] Add `buttonsEventFilter` support to `GestureRecognizer`
	- [ ] Add `buttonsEventFilter` support to `DoubleTapGestureRecognizer`
	- [ ] Add `buttonsEventFilter` support to `SerialTapGestureRecognizer`
	- [ ] Add `buttonsEventFilter` support to `MultiTapGestureRecognizer`
	- [ ] Add `buttonsEventFilter` support to `OneSequenceGestureRecognizer`
		- [ ] Add `buttonsEventFilter` support to `ScaleGestureRecognizer`
		- [ ] Add `buttonsEventFilter` support to `EagerGestureRecognizer`
		- [ ] Add `buttonsEventFilter` support to `ForcePressGestureRecognizer`
		- [ ] Add `buttonsEventFilter` support to `DragGestureRecognizer`
			- [ ] Add `buttonsEventFilter` support to `PanGestureRecognizer`
			- [ ] Add `buttonsEventFilter` support to `HorizontalDragGestureRecognizer`
			- [ ] Add `buttonsEventFilter` support to `VerticalDragGestureRecognizer`
		- [ ] Add `buttonsEventFilter` support to `PrimaryPointerGestureRecognizer`
			- [ ] Add `buttonsEventFilter` support to `LongPressGestureRecognizer`
			- [ ] Add `buttonsEventFilter` support to `BaseTapGestureRecognizer`
				- [ ] Add `buttonsEventFilter` support to `TapGestureRecognizer`

Renzo-Olivares commented:
I'm also still unsure about the name. What do you think about `supportedButtonsFilter`? or would that be confusing with `supportedDevices`? Maybe `allowedButtonsFilter`? `buttonsEventFilter` seems a little confusing to me, it sounds like there is some `ButtonEvent` we are referring to but we are referring to `PointerEvent.buttons`.

bernaferrari commented on packages/flutter/lib/src/widgets/drag_target.dart:
```diff
@@ -179,6 +179,7 @@ class Draggable<T extends Object> extends StatefulWidget {
     this.ignoringFeedbackPointer = true,
     this.rootOverlay = false,
     this.hitTestBehavior = HitTestBehavior.deferToChild,
+    this.buttonsEventFilter,
```
We could just accept `PointerEvent event` instead of `int buttons`, which is fine. But might be too granular, as we are almost making `isPointerAllowed` a parameter.

a-wallen commented:
> GestureRecognizer buttonsEventFilter API support project ☔
> 
> Add buttonsEventFilter support to GestureRecognizer
> Add buttonsEventFilter support to DoubleTapGestureRecognizer
> Add buttonsEventFilter support to SerialTapGestureRecognizer
> Add buttonsEventFilter support to MultiTapGestureRecognizer
> Add buttonsEventFilter support to OneSequenceGestureRecognizer
> Add buttonsEventFilter support to ScaleGestureRecognizer
> Add buttonsEventFilter support to EagerGestureRecognizer
> Add buttonsEventFilter support to ForcePressGestureRecognizer
> Add buttonsEventFilter support to DragGestureRecognizer
> Add buttonsEventFilter support to PanGestureRecognizer
> Add buttonsEventFilter support to HorizontalDragGestureRecognizer
> Add buttonsEventFilter support to VerticalDragGestureRecognizer
> Add buttonsEventFilter support to PrimaryPointerGestureRecognizer
> Add buttonsEventFilter support to LongPressGestureRecognizer
> Add buttonsEventFilter support to BaseTapGestureRecognizer
> Add buttonsEventFilter support to TapGestureRecognizer

@Renzo-Olivares we discussed landing these in separate PRs versus landing all of these in one pull request during triage. We think that it would be more efficient if @bernaferrari landed all of these changes in one pull request.

@bernaferrari feel free to chime in and let us know what you think.

Piinks commented:
I agree with @Renzo-Olivares & @a-wallen regarding making this consistent across all of those classes. I think I raised this point earlier as well.
That would only involve piping it through as a super parameter to all of the subclasses like we did in https://github.com/flutter/flutter/pull/81858

Adding parameters to classes ad hoc creates inconsistencies in the framework. While the request is for support in this particular widget, how it relates to other widgets in the framework should be accounted for. This is embracing the yak shave. :)

bernaferrari commented:
Could you check what's failing in Google testing and if it's a big deal? Else the bigger change might trigger more things.

goderbauer commented:
> Could you check what's failing in Google testing and if it's a big deal? Else the bigger change might trigger more things.

You'll have to rebase this PR to the latest master to kick off another "google testing" run. Can you please do that?

bernaferrari committed:
DragTarget part 1.

[WIP] Change GestureRecognizer. Sorry.

[WIP] Move from GestureRecognizer to MultiDragGestureRecognizer.

Make it a `Set<int>?`

Get bitwise operations working.

Fix test. Rename to allowedInputPointers.

Convert into a builder.

Improve code with default funciton.

Refactor everything again.

Rename to buttonEventFilter.

Use static function.

Fix analyzer.

Fix private reference.

Use // in private method.

bernaferrari committed:
Fix Renzo request.

bernaferrari committed:
Add `allowedButtonsFilter` everywhere.

bernaferrari committed:
Refactor monoDrag for multi pointer support.

bernaferrari commented:
Hey! Sorry for the end of the year delay! 😅 Tests are ok now.

So, I just did what you asked me to. I added `super.allowedButtonsFilter` everywhere. This creates a few "inconsistencies", but so far so good.

For example, `LongPressGestureRecognizer` independently tracks primary/secundary/tertiary buttons. Theoretically, allowedButtonsFilter could handle that, and allow for more complex customization, but these are sent as callback (such as `onSecondaryLongPressDown`). So unless we convert the 3 variations of each into a single and expose to the user (like `onLongPressDown(int mouseButtons)`), I don't think it would work. Given it is used in Editable, it is unrealistic to expect different longPress configurations. It would be better for it to be more versatile, but I think we would make it too complex for what it needs to be. 

On DragGestureRecognizer, on the other hand, I replaced where it tracked `kPrimaryButton` with support for anything. So now it is more versatile than ever.

bernaferrari committed:
Fix tests?

bernaferrari committed:
Change default to always true.

bernaferrari commented:
- What should I do with DoubleTapGestureRecognizer? It is primary button only right now.
- Should we do anything with LongPressGestureRecognizer? I don't think so.
I changed the default `allowedButtonsFilter` to allow all, and on DragGestureRecognizer (slider) and MultiDragGestureRecognizer (drag and drop) I made the default `kPrimaryButton` only.

Renzo-Olivares reviewed (commented):
Thank you for making all these changes @bernaferrari !! It looks great!

* What should I do with DoubleTapGestureRecognizer? It is primary button only right now.
I think it's okay here to add `allowedButtonsFilter` and have it default to primary button. No complications come to mind immediately by letting `DoubleTapGestureRecognizer` accept different buttons since it checks if an incoming `PointerDownEvent` has the same buttons as the previous to verify the double tap. Was there anything that felt concerning to you?

* Should we do anything with LongPressGestureRecognizer? I don't think so.
I am a bit unsure for this one. Couldn't we just add `allowedButtonsFilter` and set the default to accept primary/secondary/tertiary buttons? I'm unsure why we have to pipe the buttons to the callbacks. I understand the buttons are tracked by the recognizer in `_initialButtons` but that is first tracked in `handlePrimaryPointer` which is called after `isPointerAllowed`. I think this means that we can assume the `_initialButtons` saved is still valid.

bernaferrari commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -98,6 +105,29 @@ abstract class GestureRecognizer extends GestureArenaMember with DiagnosticableT
+  /// When value is `(int buttons) => false`, allow no interactions.
+  /// When value is `(int buttons) => true`, allow all interactions.
+  ///
+  /// Defaults to primary button only.
```
Yes! Thanks. The "accept all buttons" change was the last one.

bernaferrari commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -98,6 +105,29 @@ abstract class GestureRecognizer extends GestureArenaMember with DiagnosticableT
   /// tracked and recognized.
   final Set<PointerDeviceKind>? _supportedDevices;
 
+    /// {@template flutter.gestures.multidrag._allowedButtonsFilter}
```
Thanks! Fixed.

bernaferrari commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -48,6 +48,11 @@ enum DragStartBehavior {
 }
 
+/// Signature for `allowedButtonsFilter` in [GestureRecognizer].
+/// Used to filter the input buttons of the pointer event.
```
Done :)

bernaferrari commented on packages/flutter/lib/src/gestures/recognizer.dart:
```diff
@@ -187,7 +217,9 @@ abstract class GestureRecognizer extends GestureArenaMember with DiagnosticableT
   bool isPointerAllowed(PointerDownEvent event) {
     // Currently, it only checks for device kind. But in the future we could check
```
Thanks, fixed.

bernaferrari committed:
Fix PR comments.

bernaferrari committed:
Completely refactor long press.

bernaferrari committed:
Revert "Completely refactor long press."

This reverts commit 5038e8603e250e8c928b0f1754fb794b7b75738b.

bernaferrari commented:
`DoubleTapGestureRecognizer` is going to call `super.isPointerAllowed(event)` which triggers the change we are making. By default it accept all, so no changes. But you can do weird things, because `allowedButtonsFilter` can block different buttons, so, by default primary -> secondary -> primary, not a double tap because there was a middle tap of different button. Now, if the user changes allowedButtonFilter, primary -> secondary (ignored) -> primary -> double tap is a success.

We could further relax `DoubleTapGestureRecognizer`, like here. But I think it is ok in its current situation:
![image](https://user-images.githubusercontent.com/351125/210870325-9b49a514-9a47-41de-a7ff-3c9c2357887f.png)

Regarding `LongPressGestureRecognizer`,
a) by default it accepts a single button, and it only really works if a callback is enabled. So the only thing `allowedButtonsFilter` could do is make it not work at all, like a disable method.
b) However... `allowedButtonsFilter` could be by default "button == primary || button == secondary || button == tertiary", and, relaxing further, situations like `buttons = primary | secondary` could be allowed. I just did this. The change was relatively big and I would like to know how you feel about it.

bernaferrari committed:
Add forgotten class.

Renzo-Olivares commented:
> `DoubleTapGestureRecognizer` is going to call `super.isPointerAllowed(event)` which triggers the change we are making. By default it accept all, so no changes. But you can do weird things, because `allowedButtonsFilter` can block different buttons, so, by default primary -> secondary -> primary, not a double tap because there was a middle tap of different button. Now, if the user changes allowedButtonFilter, primary -> secondary (ignored) -> primary -> double tap is a success.
> 
> We could further relax `DoubleTapGestureRecognizer`, like here. But I think it is ok in its current situation: ![image](https://user-images.githubusercontent.com/351125/210870325-9b49a514-9a47-41de-a7ff-3c9c2357887f.png)
> 
> Regarding `LongPressGestureRecognizer`, a) by default it accepts a single button, and it only really works if a callback is enabled. So the only thing `allowedButtonsFilter` could do is make it not work at all, like a disable method. b) However... `allowedButtonsFilter` could be by default "button == primary || button == secondary || button == tertiary", and, relaxing further, situations like `buttons = primary | secondary` could be allowed. I just did this ([5038e86](https://github.com/flutter/flutter/commit/5038e8603e250e8c928b0f1754fb794b7b75738b)) and reverted 😭. The second gesture will cancel the first (you'll never get primary and secondary click at the same time, at least with a mouse), so it will only ever work with a single gesture. I'll keep `a)`.

Thank you for clarifying all of that! In the case of `DoubleTapGestureRecognizer` it sounds like we may want to leave it as is then since it sounds like it always needs to accept all buttons to prevent weird behaviors. Though I do worry a bit that not adding `allowedButtonsFilter` to some `GestureRecognizers` may lead to a inconsistent experience for a developer as they might expect `DoubleTapGestureRecognizer` to have `allowedButtonsFilter` like other `GestureRecognizers` do. @Piinks @a-wallen what are your thoughts on this?

@bernaferrari what do you mean by relaxing? I can't tell the difference in the code.

Option `a)` also sounds like a good idea to me given the setup of `LongPressGestureRecognizer`. I'm not fully understanding why we couldn't keep the default to `button == primary || button == secondary || button == tertiary` vs `all buttons`. Would `allowedButtonsFilter` still work as a disable in this case?

bernaferrari commented:
> Though I do worry a bit that not adding allowedButtonsFilter to some GestureRecognizers may lead to a inconsistent experience for a developer

We can add it, it will probably not help anyone, but also won't hurt anyone.


> I'm not fully understanding why we couldn't keep the default to button == primary || button == secondary || button == tertiary vs all buttons. Would allowedButtonsFilter still work as a disable in this case?

Sure, I can make this. AllowedButtonsFilter would disable it, but if there is no callback, it also auto disables it. The logic behind LongPress is "if mouseButton matches callback, allow the pointer". You could have the callback and use allowedButtonsFilter to cancel it.

<img width="819" alt="image" src="https://user-images.githubusercontent.com/351125/211094354-cecf9823-d754-4cc0-9625-c9c242aa10ec.png">

bernaferrari committed:
Add default value to LongPress

a-wallen commented:
> But you can do weird things, because allowedButtonsFilter can block different buttons, so, by default primary -> secondary -> primary, not a double tap because there was a middle tap of different button. Now, if the user changes allowedButtonFilter, primary -> secondary (ignored) -> primary -> double tap is a success.

I was under the impression that the double tap recognizer would start listening from its initial state if a button was not allowed. Can you explain why the second button is ignored and the double tap recognizer is still waiting for the second tap?

With that explanation, we'll have sufficient documentation if we end up adding `allowedButtonsFilter` to the double-tap recognizer.

a-wallen commented:
> isPointerAllowed is the first thing that is executed, so if it is not allowed, it is completely ignored. The other functions that detect if current buttons == initial buttons, doSomething, else doSomethingElse won't run if isPointerAllowed returns false. Right now it accepts a button it shouldn't accept and delegates the "should we accept it?" decision for later. I guess, I could try to move this decision earlier. I could try to do something like if (super.isPoinerAllowed == false, resetState()). Would you be happy with this?

If 
1. secondary is not allowed
2. `primary -> secondary -> primary` does not register as a double tap
then yes, I think that we could add the `allowedButtonsFilter` as an API to the double-tap recognizer.

bernaferrari committed:
Refactor doubleTap.

bernaferrari commented:
Fixed with https://github.com/flutter/flutter/pull/111852/commits/40dd6b4e56431273bfc09f7b79bcbdbe3c17af62 😛

a-wallen commented:
Thanks, it looks fantastic 👍

bernaferrari commented:
That's it for now, I guess. If you want to re-review @Renzo-Olivares @moffatman or approve, feel free.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/multitap.dart:
```diff
@@ -215,7 +216,13 @@ class DoubleTapGestureRecognizer extends GestureRecognizer {
-    return super.isPointerAllowed(event);
+
+    // If second tap is not allowed, reset the state.
+    final bool isPointerAllowed = super.isPointerAllowed(event);
```
With this change, should we also remove the switch case above that checks for `kPrimaryButton` but keep the check for callbacks? That would let `DoubleTapGestureRecognizer` work for any `button` and `allowedButtonsFilter` can be used to filter that. We might want to set the default for `DoubleTapGestureRecognizer.allowedButtonsFilter` to be `kPrimaryButton` as that the recognizer only previously accepted `kPrimaryButton` as its default behavior.

bernaferrari replied:
Sure, I'm going to try. I hate the tests 😭.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/long_press.dart:
```diff
@@ -258,6 +258,7 @@ class LongPressGestureRecognizer extends PrimaryPointerGestureRecognizer {
     super.kind,
     super.supportedDevices,
     super.debugOwner,
+    super.allowedButtonsFilter = _defaultButtonAcceptBehavior,
```
Should we add some documentation to this class indicating that the `allowedButtonsFilter` does not give this recognizer the ability to recognize any buttons beyond the ones it already accepts (kPrimaryButton, kSecondaryButton, kTertiaryButton)? 

I point this out because something like `DragGestureRecognizer` which does not have *secondary *tertiary callbacks can use `allowedButtonsFilter` to accept a drag on any button. But `LongPressGestureRecognizer` cannot because it is setup in a way that it has *secondary and *tertiary callbacks. This is also true for `TapGestureRecognizer` I believe.

bernaferrari replied:
Here it goes.

Renzo-Olivares commented:
Thanks for being very patient and making all these changes so quickly @bernaferrari!!

bernaferrari committed:
Relax double tap.

bernaferrari committed:
Write comment in LongPress.

Renzo-Olivares commented on packages/flutter/lib/src/gestures/long_press.dart:
```diff
@@ -247,6 +247,14 @@ class LongPressGestureRecognizer extends PrimaryPointerGestureRecognizer {
   /// The [duration] argument can be used to overwrite the default duration
   /// after which the long press will be recognized.
   ///
+  /// The [allowedButtonsFilter] argument only gives this recognizer the
```
Nit: I think you can make this a template and reuse it in the `TapGestureRecognizer`.

Renzo-Olivares reviewed (approved):
LGTM!! Thanks again for all your patience and the quick changes!!

bernaferrari committed:
Use template.

Piinks reviewed (approved):
![Flutter_LGTM](https://user-images.githubusercontent.com/16964204/169392212-02bb74fe-c6b9-4eb7-aa62-6ddb43a70d5b.png)
Thank you for your dedication to this PR!
